packages/qunit-dom/package.json

```diff
@@ -16,16 +16,16 @@
   "type": "module",
   "exports": {
     ".": {
-      "types": "./dist/addon-test-support/qunit-dom-modules.d.ts",
-      "default": "./dist/addon-test-support/index.js"
+      "types": "./dist/es/qunit-dom.d.ts",
+      "default": "./dist/es/index.js"
```

Comment:
Do we need this paths change too?
I think we should not change them if we don't need to.
Especially since the issue they originally attempted to solve was fixed with the configuration for 'iife' output.

Reply:
I can split it out if you want, but I think it helps immensely with understanding

Reply:
Overall I'm currently struggling to begin treating qunit-dom as a JS lib rather than an addon.
I agree with the changes though, guess it makes sense to rename that, since as you've mentioned previously there's no "emberism" anymore.

I'd like to ask is an `es` directory a general practice in the JS land?
I think it makes sense but was wondering if this is how it's usually done.

At this point, I believe we should also implement the convention to add an `.esm` extension to the file instead so
`/dist/es/index.js` -> `/dist/index.esm.js`
Though I've no real hard opinion on this, just wondering what'd be the most common and appropriate approach here. I've looked at outputs of a few libraries and interestingly each one did it differently :joy: 
One common approach that some of them have is `.esm.js` naming convention and ideally keep things flat.

I don't know which is best and what'd be the pros and cons of each one.
Personally I think I like what `popper.js` does best; they basically say "here's the default way to consume the package" while also including the same assets in a specific directory as a failsafe.
And then also in case of `photoswipe` types are also in a directory of their own.

At the end of the day, users will by default use whatever we declare in the package.json but I'd like to decide here what approach should we take and why.

What do you think?

Examples:
https://unpkg.com/browse/vue@3.3.4/dist/
- Vue seems to have dropped any kind of additional nesting and all files are kept in the `/dist/` and distinguishes build types based on extensions e.g. `.esm.js`, `.cjs.js`

https://unpkg.com/browse/react-dom@18.2.0/
- might not be exactly applicable as it's AFAIK a javascript that runs only in node environment, still leaving this here

https://www.unpkg.com/browse/photoswipe@5.4.2/dist/
- seems to have "default" `esm` export in the `/dist/` directly with additional file extension `.esm.js` 

https://unpkg.com/browse/popper.js@1.12.8/dist/
- seems to have both "default" `esm` export in the `/dist/` and also the same thing copied to `/dist/esm/`

Reply:
> At this point, I believe we should also implement the convention to add an .esm extension to the file instead so

there's no need because the package.json declares `type=module`, so "it's known" that all `.js` files are ESM.

the IIFE is the odd one out, because IIFEs have no place in node/bundler land. They only make sense in old school CDN usage. And even, then, we can use ESM in CDNs now, for example:
```html
Note how qunit-dom has no special treatment to the path (other than targeting the pre-release version), because we end up hitting the "exports" entry in the package.json, which ESM-based CDNs understand and respect 🥳

<script type="importmap">
  {
    "imports": {
      "qunit": "https://esm.sh/qunit",
      "qunit-dom": "https://esm.sh/qunit-dom@3.0.0-rc.1"
    }
  }
</script>
<link rel="stylesheet" href="https://esm.sh/qunit@2.20.0/qunit/qunit.css">
<script type="module">
  import * as QUnit from 'qunit';
  import { setup } from 'qunit-dom';

  setup(QUnit.assert);

</script>
```

So, _technically_ the legacy IIFE build isn't needed at all, folks wanted to use native ESM in browsers via CDN.

Here is a demonstration of how we can use `qunit-dom` on jsbin:
https://jsbin.com/fipayiy/1/edit?html,output
![image](https://github.com/mainmatter/qunit-dom/assets/199018/1b9aa077-8e38-4d35-919c-8b1c35ea5cac)


> I'd like to ask is an es directory a general practice in the JS land?

I wouldn't say there are general practices, but directories per format _are_ quite common, and all the glimmer packages do it as well.
![image](https://github.com/mainmatter/qunit-dom/assets/199018/99124272-c8ea-4db9-a048-6c4f7fbd08a4)
(tho, AMD really has no place in modern packages)


> I've looked at outputs of a few libraries and interestingly each one did it differently 😂

yeah, I mean, ultimately, it's up to each project to do what works best for them.
The package.json and exports / imports / ESM rules are what we all abide by, and beyond that, anything is fair game.

> "here's the default way to consume the package" while also including the same assets in a specific directory as a failsafe.

the whole story of how to consume a package is defined by the `exports`.
for example, if we wanted to also support commonjs (which doesn't make sense for a browser library, such as qunit-dom), we'd do:
```
"exports": {
  "*": {
    "types": "... types must be first ...",
    "require": " path for cjs-only file ",
    "import": " path for esm-only file"
  }
}
```
when using `exports.*.default`, you say all types of resolving use "this file" (the default) -- so it ends up working for cjs and esm.

`main` / `module` / `browser` fields in the package.json aren't really used anymore for browser packages.
`main` still has some usage in node libraries though.



> Vue seems to have dropped any kind of additional nesting and all files are kept in the /dist/ and distinguishes build types based on extensions

this is easiest achieved when the build tools allow for it.
since we're essentially running rollup twice, separate target directories has

Reply:
That's a blog post entry right there :open_mouth: 
Thanks a lot for this!

I thought I've had a seemingly innocent question going along these lines "How qunit-dom's build artifacts map to the general state of JS land?" and I received this gold in response :pray: